### PR TITLE
Create a Validate pkg

### PR DESCRIPTION
**Pull request by ArangoGutierrez**

### CODE REVIEW

**Comment by ArangoGutierrez** [comment]
oh I forgot unit tests at master 😓

File: pkg/apis/nfd/validate/validate.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+	nfdv1alpha1 "sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/v1alpha1"
+)
+
+var (
+	// Default error message for invalid label/annotation keys
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+)
+
+func Namespace(labelNs string, wildcardDeniedNs, normalDeniedNs, extraLabelNs map[string]struct{}) error {
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:35]: I'd probably keep this in nfd-master as it feels specific to it

File: pkg/apis/nfd/validate/validate.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+			return fmt.Errorf("namespace %q is not allowed", labelNs)
+		}
+	}
+	if _, ok := extraLabelNs[labelNs]; !ok {
+		return fmt.Errorf("namespace %q is not allowed", labelNs)
+	}
+	return nil
+}
+
+func Label(key, value string, features *nfdv1alpha1.Features) (string, error) {
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:52]: nit: Add "docstring" comments to the functions

**Review by marquiz — COMMENTED** [review]
Thanks @ArangoGutierrez for the PR. Splitting this out of #1446 makes for a lot easier review. I'd keep this patch validation-only and have the dynamic value expansion of labels inside nfd-master

File: pkg/apis/nfd/validate/validate.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+var (
+	// Default error message for invalid label/annotation keys
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+)
+
+// Label validates a label key and value and returns an error if the key or
+// value is invalid.
+func Label(key, value string, features *nfdv1alpha1.Features) error {
+
```
Comment (AhmedGrati) [inline pkg/apis/nfd/validate/validate.go:37]: nit: extra empty line

File: pkg/apis/nfd/validate/validate.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+func Taint(taint *corev1.Taint) error {
+	ns, _ := splitNs(taint.Key)
+	// if ns is empty, it means that the rule is not allowed
+	if ns == "" {
+		return fmt.Errorf("unprefixed keys are disallowed")
+	}
+	if ns != nfdv1alpha1.TaintNs && !strings.HasSuffix(ns, nfdv1alpha1.TaintSubNsSuffix) {
+		// If the namespace is denied, rule is invalid
+		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
+			return fmt.Errorf("namespace %v is not allowed", ns)
```
Comment (AhmedGrati) [inline pkg/apis/nfd/validate/validate.go:82]: I think we can use `ErrNSNotAllowed` error. wdyt?

File: pkg/apis/nfd/validate/validate.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		}
+	}
+	// Validate taint effect is not empty
+	if taint.Effect == "" {
+		return fmt.Errorf("taint effect cannot be empty")
+	}
+	// Validate effect to be only one of NoSchedule, PreferNoSchedule or NoExecute
+	if taint.Effect != "NoSchedule" && taint.Effect != "PreferNoSchedule" && taint.Effect != "NoExecute" {
```
Comment (AhmedGrati) [inline pkg/apis/nfd/validate/validate.go:90]: I think we can use taints that are already defined in the `corev1` package of Kubernetes. e.g. `taint.Effect != corev1.TaintEffectNoSchedule`

**Comment by ArangoGutierrez** [comment]
@AhmedGrati  @marquiz  PTAL

**Comment by ArangoGutierrez** [comment]
@marquiz @AhmedGrati Rebased, ready for another round

**Review by AhmedGrati — APPROVED** [review]

**Review by AhmedGrati — COMMENTED** [review]
@ArangoGutierrez I think it would be a great addition if you can add some unit tests for the `validate` functions. wdyt?

File: Makefile [hunk 6]
```diff
@@ -191,14 +191,14 @@ e2e-test:
 	    -kubeconfig=$(KUBECONFIG) \
 	    -nfd.e2e-config=$(E2E_TEST_CONFIG) \
 	    -nfd.pull-if-not-present=$(E2E_PULL_IF_NOT_PRESENT) \
-	    -ginkgo.focus="\[kubernetes-sigs\]" \
+	    -ginkgo.focus="\[kubernetes-sigs\]NFD master and worker" \
```
Comment (marquiz) [inline Makefile:194]: Don't change this. We want to run also other nfd e2e tests...

File: pkg/nfd-master/nfd-master.go [hunk 7]
```diff
@@ -1447,19 +1419,27 @@ func (m *nfdMaster) filterFeatureAnnotations(annotations map[string]string) map[
...
+			klog.ErrorS(err, "ignoring annotation", "annotationKey", annotation, "annotationValue", value)
+			continue
 		}
 
 		outAnnotations[annotation] = value
 	}
 	return outAnnotations
 }
+
+func isNamespaceDenied(labelNs string, wildcardDeniedNs map[string]struct{}, normalDeniedNs map[string]struct{}) bool {
```
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:1433]: nit: why was this func moved? Harder to review (and tell if smth changed) if things are moved around. If we want to do this put it into a separate commit for example to isolate the change

File: pkg/apis/nfd/validate/validate.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+	return nil
+}
+
+// ExtendedResource validates an extended resource key and value and returns an
+// error if the key or value is invalid.
+func ExtendedResource(key, value string, features *nfdv1alpha1.Features) error {
+	ns, _ := splitNs(key)
+
+	//Validate label name
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:153]: nit: label?

File: pkg/nfd-master/nfd-master.go [hunk 9]
```diff
@@ -814,7 +783,9 @@ func filterExtendedResource(name, value string, features *nfdv1alpha1.Features)
 	if ns == "" {
 		return "", fmt.Errorf("extended resource without namespace (prefix/) not allowed")
 	}
-	if ns != nfdv1alpha1.ExtendedResourceNs && !strings.HasSuffix(ns, nfdv1alpha1.ExtendedResourceSubNsSuffix) {
+
+	err := validate.ExtendedResource(name, value, features)
+	if err == validate.ErrNSNotAllowed {
```
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:788]: This logic looks b0rken 🤨 What if there's some other error, we just ignore it and continue happily(?)

We should move the check of `kubernetes.io`  and `kubernetes.io` below to the validate func

File: pkg/apis/nfd/validate/validate_test.go
Comment (marquiz) [inline pkg/apis/nfd/validate/validate_test.go:1]: When you're at it, add unit test for ERs, too

**Review by marquiz — COMMENTED** [review]
Thanks for the update. Some quick comments below

File: Makefile (hunk 6 above)
Comment (ArangoGutierrez) [reply Makefile:194]: I always forget to remove this .... we should have this dynamic with variables...  (TODO)

File: pkg/apis/nfd/validate/validate.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+	// Default error message for invalid label/annotation keys
+	ErrUnprefixedKeysNotAllowed = fmt.Errorf("unprefixed keys are disallowed")
+	// Default error for invalid taint effect
+	ErrInvalidTaintEffect = fmt.Errorf("invalid taint effect")
+	// Default error for empty taint effect
+	ErrEmptyTaintEffect = fmt.Errorf("empty taint effect")
+)
+
+func getDynamicValue(value string, features *nfdv1alpha1.Features) (string, error) {
```
Comment (AhmedGrati) [inline pkg/apis/nfd/validate/validate.go:42]: Can we add a test function for this one also?

Comment (ArangoGutierrez) [reply pkg/apis/nfd/validate/validate.go:42]: done

**Review by AhmedGrati — APPROVED** [review]
LGTM! Thanks for your efforts @ArangoGutierrez!

File: pkg/apis/nfd/validate/validate.go [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+		}
+		// And is not a denied namespace
+		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
+			return "", ErrNSNotAllowed
+		}
+	}
+
+	// Validate label value
+	if strings.HasPrefix(value, "@") {
+		dynamicValue, err := getDynamicValue(value, features)
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:85]: I'd suggest to keep this dynamic expansion of values in nfd-master. I.e. first expand the value and the call `validate.Labe()` to check if everything looks good. This would also make the `validate` package/API more uniform, i.e. all validate functions return just an error, not a value plus an error (and also as input you don't need to give Features).

File: pkg/apis/nfd/validate/validate.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+	if err := k8svalidation.IsQualifiedName(key); len(err) > 0 {
+		return "", fmt.Errorf("invalid label key %q: %s", key, strings.Join(err, "; "))
+	}
+	// Check label namespace, filter out if ns is not whitelisted
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureLabelNs && ns != nfdv1alpha1.ProfileLabelNs &&
+		!strings.HasSuffix(ns, nfdv1alpha1.FeatureLabelSubNsSuffix) && !strings.HasSuffix(ns, nfdv1alpha1.ProfileLabelSubNsSuffix) {
+		// And is not empty
+		if ns == "" {
+			klog.ErrorS(fmt.Errorf("labels without namespace (prefix/) not allowed"), fmt.Sprintf("Ignoring annotation %s", key))
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:74]: I'd suggest to drop the klog entries from the validate package. Just return an error and log in the consumer end

```suggestion
```

File: pkg/apis/nfd/validate/validate.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	// Validate the annotation key ande value
+	if err := k8svalidation.IsQualifiedName(key); len(err) > 0 {
+		return fmt.Errorf("invalid annotation key %q: %s", key, strings.Join(err, "; "))
+	}
+
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureAnnotationNs && !strings.HasSuffix(ns, nfdv1alpha1.FeatureAnnotationSubNsSuffix) {
+		// And is not empty
+		if ns == "" {
+			klog.ErrorS(fmt.Errorf("annotations without namespace (prefix/) not allowed"), fmt.Sprintf("Ignoring annotation %s", key))
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:112]:
```suggestion
```

File: pkg/apis/nfd/validate/validate.go [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+	k8svalidation "k8s.io/apimachinery/pkg/util/validation"
+
+	nfdv1alpha1 "sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/v1alpha1"
+)
+
+var (
+	// Default error message for invalid label/annotation keys
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+	// Default error message for invalid label/annotation keys (untyped string)
+	ErrUnprefixedKeysNotAllowed = fmt.Errorf("namespace without prefix/ not allowed")
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:34]: Namespace without prefix?? Why you changed this?

File: pkg/apis/nfd/validate/validate.go [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+	}
+	// Check label namespace, filter out if ns is not whitelisted
+	ns, _ := splitNs(key)
+	// And is not empty
+	if ns == "" {
+		return ErrUnprefixedKeysNotAllowed
+	}
+	if ns != nfdv1alpha1.FeatureLabelNs && ns != nfdv1alpha1.ProfileLabelNs &&
+		!strings.HasSuffix(ns, nfdv1alpha1.FeatureLabelSubNsSuffix) && !strings.HasSuffix(ns, nfdv1alpha1.ProfileLabelSubNsSuffix) {
+		return ErrNSNotAllowed
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:56]: Hmm, I think we should deny `kubernetes.io` here but allow others, as that's the default setup. Similar to what we do for annotations and ERs and taints below.

```suggestion
		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
			return ErrNSNotAllowed
		}
```

File: pkg/apis/nfd/validate/validate.go [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+	if err := k8svalidation.IsQualifiedName(key); len(err) > 0 {
+		return fmt.Errorf("invalid annotation key %q: %s", key, strings.Join(err, "; "))
+	}
+
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureAnnotationNs && !strings.HasSuffix(ns, nfdv1alpha1.FeatureAnnotationSubNsSuffix) {
+		// And is not empty
+		if ns == "" {
+			return ErrUnprefixedKeysNotAllowed
+		}
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:80]: nit:

Move the check for empty ns out of the other if. Makes this consistent with Label and ER validation funcs

```suggestion
	if ns == "" {
		return ErrUnprefixedKeysNotAllowed
	}

	if ns != nfdv1alpha1.FeatureAnnotationNs && !strings.HasSuffix(ns, nfdv1alpha1.FeatureAnnotationSubNsSuffix) {
```

File: pkg/apis/nfd/validate/validate.go [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+// Taint validates a taint key and value and returns an error if the key or
+// value is invalid.
+func Taint(taint *corev1.Taint) error {
+	ns, _ := splitNs(taint.Key)
+	if ns != nfdv1alpha1.TaintNs && !strings.HasSuffix(ns, nfdv1alpha1.TaintSubNsSuffix) {
+		// And is not empty
+		if ns == "" {
+			return ErrUnprefixedKeysNotAllowed
+		}
```
Comment (marquiz) [inline pkg/apis/nfd/validate/validate.go:103]: nit
```suggestion
	if ns == "" {
		return ErrUnprefixedKeysNotAllowed
	}
	if ns != nfdv1alpha1.TaintNs && !strings.HasSuffix(ns, nfdv1alpha1.TaintSubNsSuffix) {
```

File: pkg/nfd-master/nfd-master.go [hunk 18]
```diff
@@ -553,49 +552,38 @@ func (m *nfdMaster) filterFeatureLabels(labels Labels, features *nfdv1alpha1.Fea
...
-	// Check label namespace, filter out if ns is not whitelisted
+	// Validate
 	ns, base := splitNs(name)
-	if ns == "" {
-		return "", fmt.Errorf("labels without namespace (prefix/) not allowed")
-	}
-	if ns != nfdv1alpha1.FeatureLabelNs && ns != nfdv1alpha1.ProfileLabelNs &&
-		!strings.HasSuffix(ns, nfdv1alpha1.FeatureLabelSubNsSuffix) && !strings.HasSuffix(ns, nfdv1alpha1.ProfileLabelSubNsSuffix) {
+	err := validate.Label(name, filteredValue)
+	if err == validate.ErrNSNotAllowed {
```
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:570]: With the change I suggested for validate.go I'd modify this one:

```suggestion
	if err == validate.ErrNSNotAllowed || isNamespaceDenied(ns, m.deniedNs.wildcard, m.deniedNs.normal) {
```

This would change the behavior slightly so that it would become possible to deny the default `feature.node.kubernetes.io` and `profile.node.kubernetes.io` but I don't believe that would be a breaking anyone. Just document it in `docs/reference/master-configuration-reference.md` (basically remove the note that the default namespaces cannot be denied). WDYT?

File: pkg/apis/nfd/validate/validate.go [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+	k8sQuantity "k8s.io/apimachinery/pkg/api/resource"
+	k8svalidation "k8s.io/apimachinery/pkg/util/validation"
+
+	nfdv1alpha1 "sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/v1alpha1"
+)
+
+var (
+	// Default error message for invalid label/annotation keys
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+	// Default error message for invalid label/annotation keys (untyped string)
```
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply pkg/apis/nfd/validate/validate.go:33]: oh I was testing something, but in the end I regret it,.... forgot to delete the comment

File: docs/usage/customization-guide.md [hunk 20]
```diff
@@ -475,9 +475,6 @@ The namespace part (i.e. prefix) of the labels is controlled by nfd:
     [`-extra-label-ns`](../reference/master-commandline-reference.md#-extra-label-ns)
     command line flag of nfd-master.
     e.g: `nfd-master -deny-label-ns="*" -extra-label-ns=example.com`
-  - Built-in default namespaces `feature.node.kubernetes.io` and
```
[the remark this answers is not in the document]
Comment (marquiz) [reply docs/usage/customization-guide.md:478]: But there's another similar note in `docs/reference/master-configuration-reference.md`

File: scripts/test-infra/test-e2e.sh [hunk 21]
```diff
@@ -11,6 +11,7 @@ chmod a+x /usr/local/bin/kubectl
 # Configure environment
 export KUBECONFIG=`pwd`/kubeconfig
 export E2E_TEST_CONFIG=`pwd`/e2e-test-config
+export E2E_TEST_FULL_IMAGE=true
```
Comment (marquiz) [inline scripts/test-infra/test-e2e.sh:14]: Not in this PR

```suggestion
```

**Review by marquiz — APPROVED** [review]
Thanks @ArangoGutierrez for the contribution. Looks ready for me
/lgtm
